docs/arch/adr-0002-hdfs-raster-layers.md

```diff
@@ -0,0 +1,74 @@
+ 2. Key/Value look-ups
+    - Clear mapping from any K to a single block file
+    - Efficient seeks to any random value in the layer
+
+HDFS does not provide any active index management so we must carefully define a storage and indexing scheme that supports both of those cases.
+
+## Decision
+
+The design builds on an established pattern of mapping a multi-dimensional tile key to a one-dimensional index using a space filling curve (SFC). This requires definition of bounding spatial extent and resolution but provides a total ordering for our records.
+
+### MapFiles
+
+The layer will be sorted and written to multiple Hadoop MapFiles. `MapFile` consist of two files:
+ - `data` file is a `SequenceFile` of `LongWritable` and `BytesWritable` key/value pairs where the key is the SFC index and value bytes are Avro encoded `Vector[(K,V)]` where all `K`s map to the given SFC index.
+ - `index` file is a `SequenceFile` which maps a `LongWritable` in seen in `data` file to its offset at some defined `indexInterval`.
+
+When `MapFile` is open the `index` is read fully and allows fast random seeks into the `data` file.
+
+Each map file will consequently correspond to an SFC range from from first to last key stored in the file. Because the whole layer is sorted before being written we can assume that that ranges covered by the map files are exclusive.
+
+It will be important to know which SFC range each file corresponds to and to avoid creating an addition overall index file we record the value of the first SFC index stored in the map file as part of the file name.
+
+We experimented with using a bloom filter index, but it did not appear appropriate. Because each file will be restricted to be no bigger than a single HDFS block (64M/128M) the time to compute and store the bloom filter does not offer any speed improvements on per-file basis.
+
+### Single Value Queries
+
+In a single value query we are given an instance of `K` and we must produce a corresponding `V` or an error. The first step is to locate the `MapFile` which potentially contains `(K, V)` record. Because the layer records are indexed by their SFC index we map `K` to `i: Long` and determine which file contains potential match by examining the file listing and finding the file with maximum starting index that is less than equal `i`. At this point the `MapFile` must be opened and queried for the key.
+
+Because both listing all the files in the layer directory and opening a single MapFile we cache these steps. The file listing needs to be done only once per `HadoopValueReader` instance and we maintain an LRU cache of open `MapFile`s. Because SFC preserves some spatial locality of the records, geographically close records are likely to be close in SFC index, and we expect key/value queries to be geographically grouped, for instance requests from a map viewer, the `MapFile` LRU cache can be expected to have very high hit-rate.
```

Comment:
`Because both listing all the files in the layer directory and opening a single MapFile we cache these steps.` I feel like this is missing some words.